### PR TITLE
chore: add Nerdbank.GitVersioning and tool nbgv

### PR DESCRIPTION
**Pull request by Bertk**
- adds dotnet tool tool-manifest `dotnet-tools.json`
- introduce Nerdbank.Gitversioning which adds precise, semver-compatible git commit information (replaces semver.txt)
- local and CI builds use appropriate semantic version (not V1.0.0)

### CODE REVIEW

**Comment by Bertk** [comment]
CHANGELOG.md will not be committed automatically. The release.yml action has already commands which need to be enabled.
Another option is to create a PR for CHANGELOG.md update.

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -43,14 +45,16 @@ jobs:
       - name: Package release
         id: package_release
         run: |
-          VERSION=`cat semver.txt`
+          dotnet tool restore
+          VERSION=`nbgv get-version --variable Version`
+          NUGETVERSION=`nbgv get-version --variable NuGetPackageVersion`
```
Comment (mtsfoni) [inline .github/workflows/release.yml:50]: There seems to be causing a problem here. From my tests it looks that Version return a 4 segment version e.g. 3.6.0.2 while nugetversion returns a SemVer version.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -43,14 +45,16 @@ jobs:
...
           echo "##[set-output name=version;]$VERSION"
-          echo "##[set-output name=package_name;]CycloneDX.$VERSION.nupkg"
-          echo "##[set-output name=package_filename;]$OUTPUT/CycloneDX.$VERSION.nupkg"
+          echo "##[set-output name=package_name;]CycloneDX.$NUGETVERSION.nupkg"
+          echo "##[set-output name=package_filename;]$OUTPUT/CycloneDX.$NUGETVERSION.nupkg"
           REPO=cyclonedx/cyclonedx-dotnet
-          dotnet build --configuration Release /p:Version=$VERSION
-          dotnet pack CycloneDX/CycloneDX.csproj --configuration Release /p:Version=$VERSION --output $OUTPUT
+          dotnet build --configuration Release
+          dotnet pack CycloneDX/CycloneDX.csproj --configuration Release --output $OUTPUT
```
Comment (mtsfoni) [inline .github/workflows/release.yml:57]: The file with 3 semver version in the filename is created.
```
Successfully created package '/home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/nupkgs/CycloneDX.3.0.6.nupkg'.
```

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -43,14 +45,16 @@ jobs:
...
           docker build -f Dockerfile --build-arg VERSION=$VERSION -t $REPO:$VERSION -t $REPO:latest .
```
Comment (mtsfoni) [inline .github/workflows/release.yml:58]: now here i tried to read the file with a 4 segment version in the filename:

```
 [3/3] RUN dotnet tool install --global CycloneDX --version 3.0.6.2 --add-source /tmp/nupkgs &&     ln -s /root/.dotnet/tools/dotnet-CycloneDX /usr/bin/CycloneDX
```
resulting in:
```
#7 1.127 Unhandled exception: Microsoft.DotNet.Cli.NuGetPackageDownloader.NuGetPackageNotFoundException: cyclonedx::3.0.6.2 is not found in NuGet feeds https://api.nuget.org/v3/index.json;/tmp/nupkgs".
#7 1.135    at Microsoft.DotNet.Cli.NuGetPackageDownloader.NuGetPackageDownloader.GetPackageMetadataAsync(String packageIdentifier, NuGetVersion packageVersion, IEnumerable`1 sources, CancellationToken cancellationToken, Boolean includeUnlisted)
#7 1.135    at Microsoft.DotNet.Cli.NuGetPackageDownloader.NuGetPackageDownloader.GetPackageSourceAndVersion(PackageId packageId, NuGetVersion packageVersion, PackageSourceLocation packageSourceLocation, Boolean includePreview, Boolean includeUnlisted, PackageSourceMapping packageSourceMapping)
#7 1.135    at Microsoft.DotNet.Cli.NuGetPackageDownloader.NuGetPackageDownloader.DownloadPackageAsync(PackageId packageId, NuGetVersion packageVersion, PackageSourceLocation packageSourceLocation, Boolean includePreview, Boolean includeUnlisted, Nullable`1 downloadFolder, PackageSourceMapping packageSourceMapping)
#7 1.135    at Microsoft.DotNet.Cli.ToolPackage.ToolPackageDownloader.DownloadAndExtractPackage(PackageLocation packageLocation, PackageId packageId, INuGetPackageDownloader nugetPackageDownloader, String packagesRootPath, IToolPackageStore toolPackageStore, NuGetVersion packageVersion, PackageSourceLocation packageSourceLocation, Boolean includeUnlisted)
#7 1.135    at Microsoft.DotNet.Cli.ToolPackage.ToolPackageDownloader.<>c__DisplayClass8_0.<InstallPackage>b__0()
#7 1.135    at Microsoft.DotNet.Cli.TransactionalAction.Run[T](Func`1 action, Action commit, Action rollback)
#7 1.135    at Microsoft.DotNet.Tools.Tool.Install.ToolInstallGlobalOrToolPathCommand.Execute()
#7 1.135    at System.CommandLine.Invocation.InvocationPipeline.Invoke(ParseResult parseResult)
#7 1.135    at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, TimeSpan startupTime, ITelemetry telemetryClient)
#7 ERROR: process "/bin/sh -c dotnet tool install --global CycloneDX --version ${VERSION} --add-source /tmp/nupkgs &&     ln -s /root/.dotnet/tools/dotnet-CycloneDX /usr/bin/CycloneDX" did not complete successfully: exit code: 1
```

**Review by mtsfoni — CHANGES_REQUESTED** [review]
I suppose we have no need for the 4 segment version. Can you change it to only use SemVer?

Aside from that, it automatically inferred, that the next version is 3.0.6. I guess that is correct behaviour.

**Comment by Bertk** [comment]
@mtsfoni The current **_version.json_** file uses a MAJOR.MINOR.PATCH for nuget package version. The `-g83366fc6b7` will be added for branches which are not defined in "**publicReleaseRefSpec**" (typically included **master** or **main**).
image

The file version information for exe/dll will include number for commits.
image

image

The 4 digit version number for file version information is also used with old version schema.
image

File: .github/workflows/release.yml (hunk 3 above)
Comment (Bertk) [reply .github/workflows/release.yml:58]: The nuget packages has version “3.0.6” (package name: CycloneDX.3.0.6.nupkg).

File: .github/workflows/release.yml (hunk 1 above)
Comment (Bertk) [reply .github/workflows/release.yml:50]: > Aside from that, it automatically inferred, that the next version is 3.0.6. I guess that is correct behaviour.

Yes, the version for the next version shall be configured after a release using **nbgv tool** or **manually**.

image

File: .github/workflows/release.yml (hunk 2 above)
Comment (Bertk) [reply .github/workflows/release.yml:57]: The version for nuget package uses  “X.X.X” and the assembly version uses “X.X.X.X”.  This was also schema was also used before.

image

**Pull request closed without merging**